Mitigate shape mismatch issue for optimize_ddp_lazy_compile=True

BoyuanFeng opened the pull request:
Summary:
`DDP Optimizer lazy compile` does not support dynamic shape yet and may report shape mismatch error. See detail: https://github.com/pytorch/pytorch/issues/116300

The root cause is that, in lazy compile, compilation is not conducted until the first `forward` call, while the symbolic shape  information is missing at that time.

This diff detects dynamic shapes and explicitly raise an user error.

Test Plan: CI

Differential Revision: D54091447




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

yf225 commented on test/distributed/test_dynamo_distributed.py:
```diff
@@ -724,6 +724,7 @@ def opt_fn(inputs):
 
 
     @patch.object(config, "optimize_ddp", True)
+    @patch.object(config, "optimize_ddp_lazy_compile", True)
```
wonder is there a way to check that it does produce the error message that you added in this PR?

yf225 commented on test/distributed/test_dynamo_distributed.py:
```diff
@@ -749,14 +750,20 @@ def forward(self, x):
+                            "we expect input tensor shapes remain static but found dynamic shapes instead. "
+                            "Please try to a) pad the input tensor to a static shape, or b) set "
+                            "config.ddp_optimizer=False. Note that config.ddp_optimizer=False may slowdown "
+                            "the performance" in msg for msg in cm.output))
```
maybe option (c) is to try to turn off `config.optimize_ddp_lazy_compile`?

BoyuanFeng committed:
Mitigate shape mismatch issue for optimize_ddp_lazy_compile=True

Summary:
`DDP Optimizer lazy compile` does not support dynamic shape yet and may report shape mismatch error. See detail: https://github.com/pytorch/pytorch/issues/116300

The root cause is that, in lazy compile, compilation is not conducted until the first `forward` call, while the symbolic shape  information is missing at that time.

This diff detects dynamic shapes and fall back to eager model when input tensor has dynamic shapes.

Test Plan: CI

Differential Revision: D54091447

eellison reviewed (commented):
I think we should land https://github.com/pytorch/pytorch/pull/120523 instead of this PR. We can support this case without throwing.

eellison commented:
Hmm, maybe we can land this, but fundamentally there's nothing preventing us from using the strides that inductor computes from one layer of DDP Optimizer to the next. Ideally we wouldn't have a matrix of supported/unsupported different options.
